jax/interpreters/partial_eval.py

```diff
@@ -316,38 +316,11 @@ def partial_eval(self, f: lu.WrappedFun, pvals: Sequence[PartialVal],
     return jaxpr, out_pvs, consts, env_tracers
 
   def process_custom_jvp_call(self, prim, fun, jvp, tracers):
-    tracers = map(self.instantiate_const_abstracted, tracers)
-    in_avals, in_consts = unzip2(t.pval for t in tracers)  # in_consts are units
-    fun = trace_to_subjaxpr(fun, self.main, True)
-    fun, aux = partial_eval_wrapper(fun, tuple(in_avals))
-    out_flat = prim.bind(fun, jvp, *in_consts)
-    out_avals, jaxpr, env = aux()
-    out_consts, consts = split_list(out_flat, [len(out_flat)-len(jaxpr.constvars)])
-    out_pvals = map(PartialVal, zip(out_avals, out_consts))  # out_consts are units
-    env_tracers = map(self.full_raise, env)
-    out_tracers = [JaxprTracer(self, pval, None) for pval in out_pvals]
-    const_tracers = map(self.new_instantiated_const, consts)
-    in_tracers = (*const_tracers, *env_tracers, *tracers)
-    closed_jaxpr = core.ClosedJaxpr(convert_constvars_jaxpr(jaxpr), ())
-
-    @_memoize
-    def jvp_jaxpr_thunk():
-      jvp_ = trace_to_subjaxpr(jvp, self.main, True)
-      jvp_, aux = partial_eval_wrapper(jvp_, tuple(in_avals) * 2)
-      with core.new_sublevel():
-        out_flat = jvp_.call_wrapped(*(in_consts * 2))  # in_consts are units
-      out_avals, jaxpr, env = aux()
-      _, consts = split_list(out_flat, [len(out_flat)-len(jaxpr.constvars)])
-      converted_jaxpr = convert_envvars_to_constvars(jaxpr, len(env))
-      return converted_jaxpr, (*consts, *env)
-
-    eqn = new_eqn_recipe(in_tracers, out_tracers, prim.initial_style,
-                         dict(fun_jaxpr=closed_jaxpr,
-                              jvp_jaxpr_thunk=jvp_jaxpr_thunk,
-                              num_consts=len(consts) + len(env)),
-                         source_info_util.current())
-    for t in out_tracers: t.recipe = eqn
-    return out_tracers
+    # We assume partial evaluation is only performed to build linear functions,
+    # and hence we don't need to keep the custom JVP rule around anymore.
```

Comment:
Notice that we already make this assumption, e.g. in `post_process_custom_jvp_call`. We just hadn't fully leveraged it because until #3370 we needed this approach for staging.

Reply:
Actually, it turns out that we still need it for `remat`, until we upgrade to the new implementation in ad_checkpoint.py